From 630f0fd79a26c864bf09ddc295b6711f788fefd3 Mon Sep 17 00:00:00 2001
From: Alexander Van Parys <alexander.vanparys@softathome.com>
Date: Thu, 5 Jun 2025 17:29:20 +0200
Subject: [PATCH] Ubusd_proto: Remove object availability check in response
 handler

It is possible to have a situation where a response to a request via
ubus is never received.

The following procedure can be followed to reproduce this issue:
 1) Client A does a lookup for object X (provided by client B) and gets a valid  ID from ubusd
 2) Client A sends a request on object X to ubusd
 3) Ubusd forwards the request to client B
 4) Client B removes object X from ubusd
 5) Client B sends a response to the request from client A to ubusd
 6) Ubusd drops the response because object X is no longer known, causing client A to be waiting forever.

The issue can be avoided by removing the object availability check in
the ubusd response handler. The response from client B will then still
be sent to client A.

Issue: HOP-10003 [Ubusd] No response received when object is removed
---
 ubusd_proto.c | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/ubusd_proto.c b/ubusd_proto.c
index 48de9b9..d1e6799 100644
--- a/ubusd_proto.c
+++ b/ubusd_proto.c
@@ -412,20 +412,11 @@ static struct ubus_client *ubusd_get_client_by_id(uint32_t id)
 
 static int ubusd_handle_response(struct ubus_client *cl, struct ubus_msg_buf *ub, struct blob_attr **attr)
 {
-	struct ubus_object *obj;
-
 	if (!attr[UBUS_ATTR_OBJID] ||
 	    (ub->hdr.type == UBUS_MSG_STATUS && !attr[UBUS_ATTR_STATUS]) ||
 	    (ub->hdr.type == UBUS_MSG_DATA && !attr[UBUS_ATTR_DATA]))
 		goto out;
 
-	obj = ubusd_find_object(blob_get_u32(attr[UBUS_ATTR_OBJID]));
-	if (!obj)
-		goto out;
-
-	if (cl != obj->client)
-		goto out;
-
 	cl = ubusd_get_client_by_id(ub->hdr.peer);
 	if (!cl)
 		goto out;
-- 
2.34.1

